### PR TITLE
Add pre-commit.com framework hooks

### PR DESCRIPTION
**Pull request by rambo**
For some basic good practises, see the [framework for more info][precommit] but TLDR is that it handles passing
only files that are in the commit to the hooks and it's very easy to manage hooks.

[precommit]: https://pre-commit.com/

Fixed some executable flags while at it but all the whitespace normalizations and other changes that the hooks will do automatically will wait for whenever someone actually touches code so that we don't have a giant commit doing only those (as discussed in #7170 )

### CODE REVIEW

**Comment by rambo** [comment]
ping @Zen-CODE, here's the conservative version.

File: .pre-commit-config.yaml
Comment (matham) [inline .pre-commit-config.yaml:9]: I'm not sure about this one. There are sometimes you may commit to master, although it's not frequent I suppose.

File: .pre-commit-config.yaml
Comment (matham) [inline .pre-commit-config.yaml:20]: This is duplicated above.

File: .pre-commit-config.yaml
Comment (matham) [inline .pre-commit-config.yaml:32]: How does this work on Windows where git will auto convert crlf to lf before committing? Would it cause a error?

File: Makefile
Comment (matham) [inline Makefile:99]:
```suggestion
	pre-commit install -f --config=kivy/tools/pep8checker/.pre-commit-config.yaml
```

Can we please move the config file to `kivy/tools/pep8checker/.pre-commit-config.yaml` so that it's not polluting the root?

**Review by matham — CHANGES_REQUESTED** [review]
You may want to add to the [docs ](https://kivy.org/doc/stable/contribute.html#coding-style), that `--no-verify` or `-n` can be used to temporarily disable the hook.

File: .pre-commit-config.yaml
Comment (rambo) [reply .pre-commit-config.yaml:9]: You can always skip a check.Docs should still be updated with pointers to pre-commit.com documentation (because you can also say to skip single check instead of the whole hook).

File: .pre-commit-config.yaml
Comment (rambo) [reply .pre-commit-config.yaml]: removed duplicate

File: Makefile
Comment (rambo) [reply Makefile:99]: it's possible sure, but root of repo is the canonical location and the pre-commit framework has an optional post-clone -hook to check for that file and auto-install it.

File: .pre-commit-config.yaml
Comment (rambo) [reply .pre-commit-config.yaml:31]: None of my mates at work that are on windows have complained but I'll fire up a VM and doublecheck this personally. That'll probably happen tomorrow though (or maybe much later tonight is I can't sleep or something).

I'm pretty sure the lf<->crlf conversion that git does on windows is done in such way that the hook only sees the actual content going to repo (which would have been converted back to lf)

File: .pre-commit-config.yaml
Comment (matham) [reply .pre-commit-config.yaml:9]: Probably both. I'm more likely to remember to use `-n` than the specific syntax to disable a specific hook and having to remember its exact name.

Comment (matham) [reply .pre-commit-config.yaml:9]: But I guess most of the time commiting to master is not desired so I guess it's fine as long as we doc how to turn off the hook.

File: Makefile
Comment (matham) [reply Makefile:99]: Yeah, but I still would rather not pollute our project root with more config files than absolutely necessery. They accumulate quickly.

I guess we'll see what the others think.

Comment (Zen-CODE) [reply Makefile:99]: Although root is the standard place,  there are pre-existing standards for kivy. I would 2nd @matham here. In this case,  kivy/tools/pep8checker is a cleaner and more consistent place.

File: .pre-commit-config.yaml
Comment (rambo) [reply .pre-commit-config.yaml:9]: the hook name is shown when it fails so remembering it is not a problem :) Also committing/pushing directly to master should be so extraordinary as not to  matter the slightest if it's difficult, way bigger problems have come up if direct pushing to master is *actually needed* to solve an issue.

File: .pre-commit-config.yaml
Comment (rambo) [reply .pre-commit-config.yaml:31]: hmm, this is getting way more complicated than just installing git and python for windows, what kind of env are the windows people actually running for development ?? If they're on WSL or cygwin they probably do not have the CRLF conversion even enabled if not how do they even enable the pre-commit hooks (ie where do they get `make` from, for example...)

Anyway gits own documentation says the conversion is done before committing and the hook sees the to-be-committed content so this *shoud* be transparent.

File: Makefile
Comment (rambo) [reply Makefile]: Moved it to a separate directory as is does more than just pep8 and the old checker script should be removed after a deprecation period.

File: .pre-commit-config.yaml
Comment (matham) [reply .pre-commit-config.yaml]: Personally I use msys and I set `autocrlf=input`.

**Review by matham — APPROVED** [review]
Looks good, thanks. It's a pretty innocuous change, but I'll wait until after the upcoming release/tag before merging on the off-chance it causes issues.

**Pull request merged**
